include/CLI/App.hpp

```diff
@@ -1124,17 +1131,31 @@ class App {
     ///@{
 
     /// Add a subcommand. Inherits INHERITABLE and OptionDefaults, and help flag
-    App *add_subcommand(std::string subcommand_name, std::string description = "") {
-        CLI::App_p subcom(new App(description, subcommand_name, this));
-        for(const auto &subc : subcommands_)
-            if(subc->check_name(subcommand_name) || subcom->check_name(subc->name_))
-                throw OptionAlreadyAdded(subc->name_);
-        subcommands_.push_back(std::move(subcom));
-        return subcommands_.back().get();
+    App *add_subcommand(std::string subcommand_name = "", std::string description = "") {
+        CLI::App_p subcom = std::shared_ptr<App>(new App(description, subcommand_name, this));
+        auto ptr = subcom.get();
+        add_subcommand(std::move(subcom));
+        return ptr;
+    }
+
+    /// Add a previously created app as a subcommand
+    App *add_subcommand(CLI::App_p subcom) {
+        if(subcom) {
+            if(!subcom->name_.empty()) {
+                for(const auto &subc : subcommands_)
+                    if(subc->check_name(subcom->name_) || subcom->check_name(subc->name_))
+                        throw OptionAlreadyAdded(subc->name_);
+            }
+            subcom->parent_ = this;
+            subcommands_.push_back(std::move(subcom));
+        }
+        return this;
```

Comment:
I am not entirely certain what this function should return.  Right now it returns a pointer to the main app so it can be chained,  but to match the other add_subcommand functions it should return a pointer to the added subcommand.  

Not sure which is preferred here.

Reply:
It either needs to match the first one or have a different name. I don't think chaining is important here.

Reply:
ok, that sounds good.  Should it throw if the pointer if not valid or return a nullptr?

Reply:
I think what you have is fine.